File: Plugins/searchengines/flashplugin.zig
//! Minimal Flash plugin protocol shim for Zig (stdlib only).
//!
//! Speaks the wire contract from docs/plugin-protocol.md: length-prefixed
//! MessagePack over stdio (4-byte big-endian length + one value) and the
//! protocol v3 lifecycle, including the warm-catalog side (publish before
//! ready, answer `sources.snapshot` from memory). Uses raw POSIX
//! read/write on fds 0/1 and fixed buffers, so it stays insulated from
//! std.Io churn across Zig releases. Decoded strings are zero-copy slices
//! into the frame buffer, valid until the next frame.

const std = @import("std");

pub const protocol_version: i64 = 3;

pub const KV = struct { key: []const u8, value: Value };

pub const nil_value: Value = .nil;

pub const Value = union(enum) {
    nil,
    boolean: bool,
    integer: i64,
    string: []const u8,
    array: []Value,
    map: []KV,

    /// Map lookup that folds a miss to nil — chainable:
    /// `msg.field("params").field("subcommand").asString()`.
    pub fn field(self: Value, key: []const u8) Value {
        return self.get(key) orelse nil_value;
    }

    pub fn get(self: Value, key: []const u8) ?Value {
        switch (self) {
            .map => |entries| {
                for (entries) |entry| {
                    if (std.mem.eql(u8, entry.key, key)) return entry.value;
                }
                return null;
            },
            else => return null,
        }
    }

    pub fn asString(self: Value) ?[]const u8 {
        return switch (self) {
            .string => |s| s,
            else => null,
        };
    }

    pub fn asInteger(self: Value) ?i64 {
        return switch (self) {
            .integer => |n| n,
            else => null,
        };
    }
};

pub const DecodeError = error{ Malformed, OutOfMemory };

pub const Decoder = struct {
    buf: []const u8,
    pos: usize = 0,
    arena: std.mem.Allocator,

    fn byte(self: *Decoder) DecodeError!u8 {
        if (self.pos >= self.buf.len) return error.Malformed;
        const b = self.buf[self.pos];
        self.pos += 1;
        return b;
    }

    fn take(self: *Decoder, n: usize) DecodeError![]const u8 {
        if (self.pos + n > self.buf.len) return error.Malformed;
        const out = self.buf[self.pos .. self.pos + n];
        self.pos += n;
        return out;
    }

    fn beInt(self: *Decoder, comptime T: type) DecodeError!T {
        const raw = try self.take(@sizeOf(T));
        return std.mem.readInt(T, raw[0..@sizeOf(T)], .big);
    }

    pub fn decode(self: *Decoder) DecodeError!Value {
        const b = try self.byte();
        if (b == 0xc0) return .nil;
        if (b == 0xc2) return .{ .boolean = false };
        if (b == 0xc3) return .{ .boolean = true };
        if (b <= 0x7f) return .{ .integer = b };
        if (b >= 0xe0) return .{ .integer = @as(i64, @as(i8, @bitCast(b))) };
        if (b >= 0xa0 and b <= 0xbf) return .{ .string = try self.take(b & 0x1f) };
        switch (b) {
            0xd9 => return .{ .string = try self.take(try self.byte()) },
            0xda => return .{ .string = try self.take(try self.beInt(u16)) },
            0xdb => return .{ .string = try self.take(try self.beInt(u32)) },
            0xcc => return .{ .integer = try self.byte() },
            0xcd => return .{ .integer = try self.beInt(u16) },
            0xce => return .{ .integer = try self.beInt(u32) },
            0xcf => return .{ .integer = @bitCast(try self.beInt(u64)) },
            0xd0 => return .{ .integer = @as(i8, @bitCast(try self.byte())) },
            0xd1 => return .{ .integer = @as(i16, @bitCast(try self.beInt(u16))) },
            0xd2 => return .{ .integer = @as(i32, @bitCast(try self.beInt(u32))) },
            0xd3 => return .{ .integer = @bitCast(try self.beInt(u64)) },
            else => {},
        }
        var count: usize = 0;
        var is_map = false;
        if (b >= 0x80 and b <= 0x8f) {
            count = b & 0x0f;
            is_map = true;
        } else if (b == 0xde) {
            count = try self.beInt(u16);
            is_map = true;
        } else if (b == 0xdf) {
            count = try self.beInt(u32);
            is_map = true;
        } else if (b >= 0x90 and b <= 0x9f) {
            count = b & 0x0f;
        } else if (b == 0xdc) {
            count = try self.beInt(u16);
        } else if (b == 0xdd) {
            count = try self.beInt(u32);
        } else {
            return error.Malformed;
        }
        if (is_map) {
            const entries = try self.arena.alloc(KV, count);
            for (entries) |*entry| {
                const key = try self.decode();
                entry.key = key.asString() orelse return error.Malformed;
                entry.value = try self.decode();
            }
            return .{ .map = entries };
        }
        const items = try self.arena.alloc(Value, count);
        for (items) |*item| item.* = try self.decode();
        return .{ .array = items };
    }
};

/// Growable encode buffer over a fixed allocator-free window: the caller
/// supplies storage sized for the largest frame it will ever emit.
pub const Encoder = struct {
    buf: []u8,
    len: usize = 0,

    fn push(self: *Encoder, bytes: []const u8) void {
        std.debug.assert(self.len + bytes.len <= self.buf.len);
        @memcpy(self.buf[self.len .. self.len + bytes.len], bytes);
        self.len += bytes.len;
    }

    fn pushByte(self: *Encoder, b: u8) void {
        self.push(&[_]u8{b});
    }

    fn pushBe(self: *Encoder, comptime T: type, value: T) void {
        var raw: [@sizeOf(T)]u8 = undefined;
        std.mem.writeInt(T, &raw, value, .big);
        self.push(&raw);
    }

    pub fn nil(self: *Encoder) void {
        self.pushByte(0xc0);
    }

    pub fn boolean(self: *Encoder, value: bool) void {
        self.pushByte(if (value) 0xc3 else 0xc2);
    }

    pub fn integer(self: *Encoder, value: i64) void {
        if (value >= 0 and value <= 127) {
            self.pushByte(@intCast(value));
        } else if (value < 0 and value >= -32) {
            self.pushByte(@bitCast(@as(i8, @intCast(value))));
        } else {
            self.pushByte(0xd3);
            self.pushBe(u64, @bitCast(value));
        }
    }

    pub fn string(self: *Encoder, value: []const u8) void {
        if (value.len < 32) {
            self.pushByte(0xa0 | @as(u8, @intCast(value.len)));
        } else {
            self.pushByte(0xdb);
            self.pushBe(u32, @intCast(value.len));
        }
        self.push(value);
    }

    pub fn mapHead(self: *Encoder, count: usize) void {
        if (count < 16) {
            self.pushByte(0x80 | @as(u8, @intCast(count)));
        } else {
            self.pushByte(0xde);
            self.pushBe(u16, @intCast(count));
        }
    }

    pub fn arrayHead(self: *Encoder, count: usize) void {
        if (count < 16) {
            self.pushByte(0x90 | @as(u8, @intCast(count)));
        } else {
            self.pushByte(0xdc);
            self.pushBe(u16, @intCast(count));
        }
    }
};

fn writeAll(bytes: []const u8) void {
    var written: usize = 0;
    while (written < bytes.len) {
        const n = std.c.write(1, bytes.ptr + written, bytes.len - written);
        if (n <= 0) return;
        written += @intCast(n);
    }
}

pub fn sendFrame(encoder: *const Encoder) void {
    var header: [4]u8 = undefined;
    std.mem.writeInt(u32, &header, @intCast(encoder.len), .big);
    writeAll(&header);
    writeAll(encoder.buf[0..encoder.len]);
}

/// Read one length-prefixed frame into `storage`; null on EOF (host gone).
pub fn readFrame(storage: []u8) ?[]const u8 {
    var header: [4]u8 = undefined;
    if (!readExact(&header)) return null;
    const n = std.mem.readInt(u32, &header, .big);
    if (n > storage.len) return null;
    const payload = storage[0..n];
    if (!readExact(payload)) return null;
    return payload;
}

fn readExact(out: []u8) bool {
    var got: usize = 0;
    while (got < out.len) {
        const n = std.c.read(0, out.ptr + got, out.len - got);
        if (n <= 0) return false;
        got += @intCast(n);
    }
    return true;
}

/// Envelope helpers shared by every response the plugin sends.
pub fn beginResponse(encoder: *Encoder, id: Value) void {
    encoder.mapHead(3);
    encoder.string("jsonrpc");
    encoder.string("2.0");
    encoder.string("id");
    switch (id) {
        .integer => |n| encoder.integer(n),
        .string => |s| encoder.string(s),
        else => encoder.nil(),
    }
    encoder.string("result");
}

pub fn sendOk(encoder: *Encoder, id: Value) void {
    encoder.len = 0;
    beginResponse(encoder, id);
    encoder.mapHead(1);
    encoder.string("ok");
    encoder.boolean(true);
    sendFrame(encoder);
}

pub fn sendError(encoder: *Encoder, id: Value, message: []const u8) void {
    encoder.len = 0;
    beginResponse(encoder, id);
    encoder.mapHead(2);
    encoder.string("ok");
    encoder.boolean(false);
    encoder.string("error");
    encoder.string(message);
    sendFrame(encoder);
}

pub fn sendLog(encoder: *Encoder, level: []const u8, message: []const u8) void {
    encoder.len = 0;
    encoder.mapHead(3);
    encoder.string("jsonrpc");
    encoder.string("2.0");
    encoder.string("method");
    encoder.string("flash.log");
    encoder.string("params");
    encoder.mapHead(3);
    encoder.string("level");
    encoder.string(level);
    encoder.string("message");
    encoder.string(message);
    encoder.string("fields");
    encoder.mapHead(0);
    sendFrame(encoder);
}

File: Plugins/reminders/flashplugin.swift
// Minimal Flash plugin protocol shim for Swift (Foundation only).
//
// Speaks the wire contract from docs/plugin-protocol.md: length-prefixed
// MessagePack over stdio (4-byte big-endian length + one value) and the
// protocol v3 lifecycle, including the warm-catalog side (publish before
// ready, answer `sources.snapshot` from memory). Lifecycle and warm reads
// answer synchronously on the read thread; commands, resolution, and
// events run on a worker queue so a slow AppleScript can never starve the
// host's 5-second heartbeat — the same discipline the Rust SDK enforces.

import Foundation

let protocolVersion = 3

// MARK: - MessagePack (the subset the protocol needs)

enum MsgPack {
  static func encode(_ value: Any?) -> Data {
    var out = Data()
    encodeInto(value, &out)
    return out
  }

  private static func encodeInto(_ value: Any?, _ out: inout Data) {
    switch value {
    case nil, is NSNull:
      out.append(0xC0)
    case let bool as Bool:
      out.append(bool ? 0xC3 : 0xC2)
    case let int as Int:
      if int >= 0 && int <= 127 {
        out.append(UInt8(int))
      } else if int < 0 && int >= -32 {
        out.append(UInt8(bitPattern: Int8(int)))
      } else {
        out.append(0xD3)
        appendBigEndian(UInt64(bitPattern: Int64(int)), &out)
      }
    case let string as String:
      let raw = Data(string.utf8)
      if raw.count < 32 {
        out.append(0xA0 | UInt8(raw.count))
      } else {
        out.append(0xDB)
        appendBigEndian(UInt32(raw.count), &out)
      }
      out.append(raw)
    case let array as [Any?]:
      if array.count < 16 {
        out.append(0x90 | UInt8(array.count))
      } else {
        out.append(0xDC)
        appendBigEndian(UInt16(array.count), &out)
      }
      for element in array { encodeInto(element, &out) }
    case let map as [String: Any?]:
      if map.count < 16 {
        out.append(0x80 | UInt8(map.count))
      } else {
        out.append(0xDE)
        appendBigEndian(UInt16(map.count), &out)
      }
      for (key, element) in map {
        encodeInto(key, &out)
        encodeInto(element, &out)
      }
    default:
      fatalError("unencodable value: \(type(of: value))")
    }
  }

  private static func appendBigEndian<T: FixedWidthInteger>(_ value: T, _ out: inout Data) {
    withUnsafeBytes(of: value.bigEndian) { out.append(contentsOf: $0) }
  }

  static func decode(_ data: Data) -> Any? {
    var position = data.startIndex
    return decodeValue(data, &position)
  }

  private static func decodeValue(_ data: Data, _ position: inout Data.Index) -> Any? {
    let byte = data[position]
    position += 1
    switch byte {
    case 0xC0: return nil
    case 0xC2: return false
    case 0xC3: return true
    case 0x00...0x7F: return Int(byte)
    case 0xE0...0xFF: return Int(Int8(bitPattern: byte))
    case 0xA0...0xBF: return decodeString(data, &position, count: Int(byte & 0x1F))
    case 0xD9: return decodeString(data, &position, count: readInt(data, &position, width: 1))
    case 0xDA: return decodeString(data, &position, count: readInt(data, &position, width: 2))
    case 0xDB: return decodeString(data, &position, count: readInt(data, &position, width: 4))
    case 0x80...0x8F: return decodeMap(data, &position, count: Int(byte & 0x0F))
    case 0xDE: return decodeMap(data, &position, count: readInt(data, &position, width: 2))
    case 0xDF: return decodeMap(data, &position, count: readInt(data, &position, width: 4))
    case 0x90...0x9F: return decodeArray(data, &position, count: Int(byte & 0x0F))
    case 0xDC: return decodeArray(data, &position, count: readInt(data, &position, width: 2))
    case 0xDD: return decodeArray(data, &position, count: readInt(data, &position, width: 4))
    case 0xCC: return readInt(data, &position, width: 1)
    case 0xCD: return readInt(data, &position, width: 2)
    case 0xCE: return readInt(data, &position, width: 4)
    case 0xCF, 0xD3: return readInt(data, &position, width: 8)
    case 0xD0: return Int(Int8(bitPattern: UInt8(readInt(data, &position, width: 1))))
    case 0xD1: return Int(Int16(bitPattern: UInt16(readInt(data, &position, width: 2))))
    case 0xD2: return Int(Int32(bitPattern: UInt32(readInt(data, &position, width: 4))))
    case 0xCA:
      let bits = UInt32(readInt(data, &position, width: 4))
      return Double(Float(bitPattern: bits))
    case 0xCB:
      // readInt yields the raw bit pattern as a (possibly negative) Int;
      // converting through UInt64(_:) traps on any double with the sign bit
      // set — e.g. a window origin left of the primary display.
      let bits = UInt64(bitPattern: Int64(readInt(data, &position, width: 8)))
      return Double(bitPattern: bits)
    default:
      return nil
    }
  }

  private static func readInt(_ data: Data, _ position: inout Data.Index, width: Int) -> Int {
    var value = 0
    for _ in 0..<width {
      value = value << 8 | Int(data[position])
      position += 1
    }
    return value
  }

  private static func decodeString(_ data: Data, _ position: inout Data.Index, count: Int)
    -> String
  {
    let end = position + count
    let string = String(data: data[position..<end], encoding: .utf8) ?? ""
    position = end
    return string
  }

  private static func decodeMap(_ data: Data, _ position: inout Data.Index, count: Int)
    -> [String: Any?]
  {
    var out: [String: Any?] = [:]
    for _ in 0..<count {
      let key = decodeValue(data, &position) as? String ?? ""
      out[key] = decodeValue(data, &position)
    }
    return out
  }

  private static func decodeArray(_ data: Data, _ position: inout Data.Index, count: Int)
    -> [Any?]
  {
    var out: [Any?] = []
    out.reserveCapacity(count)
    for _ in 0..<count {
      out.append(decodeValue(data, &position))
    }
    return out
  }
}

// MARK: - Runtime

final class PluginRuntime {
  private let output = FileHandle.standardOutput
  private let input = FileHandle.standardInput
  private let writeLock = NSLock()
  private let warmLock = NSLock()
  private var warm: [String: [[String: Any?]]] = [:]
  private let workers = DispatchQueue(label: "plugin.workers", attributes: .concurrent)

  /// Runs once when initialize arrives, BEFORE the reply — publish the warm
  /// catalog here so `published_sources` satisfies the readiness gate.
  var onStart: (() -> Void)?
  var onEvent: ((_ name: String, _ payload: [String: Any?]) -> Void)?
  var onCommand: ((_ params: [String: Any?]) -> [String: Any?])?
  var onResolve: ((_ params: [String: Any?]) -> [String: Any?])?

  func setLocations(_ sourceID: String, _ candidates: [[String: Any?]]) {
    warmLock.lock()
    warm[sourceID] = candidates
    warmLock.unlock()
  }

  func hasLocations(_ sourceID: String) -> Bool {
    warmLock.lock()
    defer { warmLock.unlock() }
    return warm[sourceID] != nil
  }

  func log(_ level: String, _ message: String, fields: [String: String] = [:]) {
    send([
      "jsonrpc": "2.0",
      "method": "flash.log",
      "params": ["level": level, "message": message, "fields": fields] as [String: Any?],
    ])
  }

  private func send(_ object: [String: Any?]) {
    let payload = MsgPack.encode(object)
    var frame = Data()
    withUnsafeBytes(of: UInt32(payload.count).bigEndian) { frame.append(contentsOf: $0) }
    frame.append(payload)
    writeLock.lock()
    output.write(frame)
    writeLock.unlock()
  }

  private func respond(_ id: Any?, _ result: [String: Any?]) {
    send(["jsonrpc": "2.0", "id": id, "result": result])
  }

  func serve() {
    var buffer = Data()
    while true {
      let chunk = input.availableData
      if chunk.isEmpty { return }  // host closed stdin
      buffer.append(chunk)
      while buffer.count >= 4 {
        let length = buffer.prefix(4).reduce(0) { $0 << 8 | Int($1) }
        guard buffer.count >= 4 + length else { break }
        let payload = buffer.subdata(in: buffer.startIndex + 4..<buffer.startIndex + 4 + length)
        buffer.removeFirst(4 + length)
        guard let message = MsgPack.decode(payload) as? [String: Any?] else { continue }
        if !dispatch(message) { return }
      }
    }
  }

  private func dispatch(_ message: [String: Any?]) -> Bool {
    let method = message["method"] as? String
    let id = message["id"] ?? nil
    let params = message["params"] as? [String: Any?] ?? [:]
    switch method {
    case "initialize":
      guard params["protocol_version"] as? Int == protocolVersion else {
        respond(id, ["ok": false, "error": "protocol version mismatch"])
        return false
      }
      onStart?()
      warmLock.lock()
      let published = warm.keys.sorted()
      warmLock.unlock()
      respond(
        id,
        [
          "ok": true, "protocol_version": protocolVersion,
          "published_sources": published as [Any?],
        ])
      return true
    case "heartbeat":
      respond(id, ["ok": true])
      return true
    case "shutdown":
      respond(id, ["ok": true])
      return false
    case "sources.snapshot":
      warmLock.lock()
      let candidates = warm.sorted { $0.key < $1.key }.flatMap(\.value)
      warmLock.unlock()
      respond(id, ["candidates": candidates as [Any?]])
      return true
    case "command.invoke":
      workers.async { [self] in respond(id, onCommand?(params) ?? ["ok": false]) }
      return true
    case "candidate.resolve":
      workers.async { [self] in respond(id, onResolve?(params) ?? ["did_resolve": false]) }
      return true
    case "event":
      let name = params["name"] as? String ?? ""
      let payload = params["payload"] as? [String: Any?] ?? [:]
      workers.async { [self] in onEvent?(name, payload) }
      return true
    default:
      if !(id is NSNull) && id != nil {
        respond(id, ["ok": false, "error": "unsupported method \(method ?? "?")"])
      }
      return true
    }
  }
}

// MARK: - Subprocess helper

struct CommandResult {
  var ok: Bool
  var stdout: String
  var stderr: String
}

/// Bounded subprocess capture: hard timeout terminates the child, mirroring
/// the Rust SDK's run_command discipline (without the process-group kill —
/// osascript/open spawn no descendants worth chasing).
func runCommand(_ argv: [String], timeoutSeconds: Double) -> CommandResult {
  let process = Process()
  process.executableURL = URL(fileURLWithPath: argv[0])
  process.arguments = Array(argv.dropFirst())
  let out = Pipe()
  let err = Pipe()
  process.standardOutput = out
  process.standardError = err
  process.standardInput = FileHandle.nullDevice
  do {
    try process.run()
  } catch {
    return CommandResult(ok: false, stdout: "", stderr: "\(error)")
  }
  let killer = DispatchWorkItem { if process.isRunning { process.terminate() } }
  DispatchQueue.global(qos: .utility).asyncAfter(
    deadline: .now() + timeoutSeconds, execute: killer)
  let stdoutData = out.fileHandleForReading.readDataToEndOfFile()
  let stderrData = err.fileHandleForReading.readDataToEndOfFile()
  process.waitUntilExit()
  killer.cancel()
  return CommandResult(
    ok: process.terminationStatus == 0,
    stdout: String(data: stdoutData, encoding: .utf8) ?? "",
    stderr: String(data: stderrData, encoding: .utf8) ?? "")
}

func runOsascript(_ script: String, timeoutSeconds: Double) -> CommandResult {
  runCommand(["/usr/bin/osascript", "-e", script], timeoutSeconds: timeoutSeconds)
}

func applescriptQuote(_ value: String) -> String {
  "\""
    + value
    .replacingOccurrences(of: "\\", with: "\\\\")
    .replacingOccurrences(of: "\"", with: "\\\"")
    + "\""
}
